# Documentation for training in phylogenomics
 
This repository aims to serve as a starting point for learning how perform various phylogenomic analyses, especially how to estimate species trees from sequence data.

Some Background for this analysis is given including brief introductions to some computing subjects and links are provided to existing tutorials, including those compiled by people at Kew. The focus is on the generation of trees for angiosperm plants (but the methods can e used for other organisms).

If you would like to contribute content to this repository please contact a contributor at RGB Kew.

## Context
[ Ideas for content topics (very short paragraph on each):
* Why build trees with sequence data?
  * What does it tell you that taxonomy can't?
* Discuss models of evolution and tree building algorithms
* What sequences are required for the analysis - discuss 
  * requirement for orthologous loci
  * NGS techniques - e.g.'s:
    * RAD-Seq
    * target capture of gene regions
]

### Context references
A recent review paper on phylogenomics
Johnson et al target capture paper and Angios353 paper

## Workflow outline for estimating a species tree
This workflow is discussed in six sections below but consists of two major steps:
* **step 1:** assembly of gene sequences from samples
* **step 2** organising sequences into gene sets, aligning the genes then using the gene alignments to estimate a species tree

It is assumed that target capture data is being used and available in the form of Illumina short pair end reads.

### Sequencing data quality checks
* Adaptor trimming
* Quality trimming

### Choice of reference target data set
### 
### gene recovery from samples - mention skimming, enrichment somewhere
### Alignment of genes from samples and assessing them
###  Estimating gene trees and assessing them
### Estimating the species tree and assessment 

### Gene alignments and trees




---
<span style="font-size:small;">Copyright Â© 2024 The Board of Trustees of the Royal Botanic Gardens, Kew</span>



